pose_lookup/plugin.cpp

```diff
@@ -0,0 +1,158 @@
+    }
+
+    virtual pose_type get_true_pose() const override {
+	throw std::logic_error{"Not Implemented"};
+    }
+
+
+    virtual bool fast_pose_reliable() const override {
+	return true;
+    }
+
+    virtual bool true_pose_reliable() const override {
+	return false;
+    }
+
+   virtual void set_offset(const Eigen::Quaternionf& raw_o_times_offset) override{
+	std::lock_guard<std::mutex> lock {offset_mutex};
+	Eigen::Quaternionf raw_o = raw_o_times_offset * offset.inverse();
+	//std::cout << "pose_prediction: set_offset" << std::endl;
+	offset = raw_o.inverse();
+    }
+
+    Eigen::Quaternionf apply_offset(const Eigen::Quaternionf& orientation) const {
+	std::lock_guard<std::mutex> lock {offset_mutex};
+	return orientation * offset;
+    }
+    virtual fast_pose_type get_fast_pose([[maybe_unused]] time_type time) const override {
+		const time_type* estimated_vsync = _m_vsync_estimate->get_latest_ro();
+		time_type vsync;
+		if(estimated_vsync == nullptr) {
+			std::cerr << "Vsync estimation not valid yet, returning fast_pose for now()" << std::endl;
+			vsync = std::chrono::system_clock::now();
+		} else {
+			vsync = *estimated_vsync;
+		}
+
+		ullong lookup_time = std::chrono::nanoseconds(vsync - _m_start_of_time ).count() + dataset_first_time;
+		ullong  nearest_timestamp = 0;
+
+		if(lookup_time <= _m_sensor_data.begin()->first){
+			std::cerr << "Lookup time before first datum" << std::endl;
+			nearest_timestamp=_m_sensor_data.begin()->first;
+		}
+		else if (lookup_time >= _m_sensor_data.rbegin()->first){
+			std::cerr << "Lookup time after last datum" << std::endl;
+			nearest_timestamp=_m_sensor_data.rbegin()->first;
+		}
+		else{
+			std::map<ullong, sensor_types>::const_iterator _m_sensor_data_it_local;
+			_m_sensor_data_it_local = _m_sensor_data.cbegin();
+			ullong prev_timestamp = _m_sensor_data_it_local->first;
+			while ( _m_sensor_data_it_local != _m_sensor_data.end() ){
+				ullong cur_timestamp = _m_sensor_data_it_local->first;
+				if(lookup_time > cur_timestamp){
+					prev_timestamp = cur_timestamp;
+					++_m_sensor_data_it_local;
+				}
+				else{
+					if((lookup_time - prev_timestamp) >= (cur_timestamp - lookup_time) ){
+						nearest_timestamp = cur_timestamp;
+						break;
+					}
+					else{
+						nearest_timestamp = prev_timestamp;
+						break;
+					}
+				}
+			}
+			// Assert while loop "break"ed out (not normal exit)
+			assert(nearest_timestamp);
+		}
+
+
+		auto looked_up_pose = _m_sensor_data.find(nearest_timestamp)->second;
+		looked_up_pose.sensor_time = _m_start_of_time + std::chrono::nanoseconds{nearest_timestamp - dataset_first_time};
+		return fast_pose_type{
+			.pose = correct_pose(looked_up_pose),
+			.predict_computed_time = std::chrono::system_clock::now(),
+			.predict_target_time = vsync
+		};
+
+	}
+
+
+private:
+	const std::shared_ptr<switchboard> sb;
+
+	mutable Eigen::Quaternionf offset {Eigen::Quaternionf::Identity()};
+	mutable std::mutex offset_mutex;
```

Comment:
This needs to be a `shared_mutex`, like pose_predict, but also see the very first comment above.